flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/optimize/CommonSubGraphBasedOptimizer.scala

```diff
@@ -79,9 +82,23 @@ abstract class CommonSubGraphBasedOptimizer extends Optimizer {
         require(plan != null)
         plan
     }
-    expandIntermediateTableScan(optimizedPlan)
+    val expanded = expandIntermediateTableScan(optimizedPlan)
+
+    val postOptimizedPlan = postOptimize(expanded)
+
+    // Rewrite same rel object to different rel objects
+    // in order to get the correct dag (dag reuse is based on object not digest)
+    val shuttle = new SameRelObjectShuttle()
+    val relsWithoutSameObj = postOptimizedPlan.map(_.accept(shuttle))
+
+    // reuse subplan
+    val tableConfig = roots.head.getTable.unwrap(classOf[TableConfig])
+    SubplanReuser.reuseDuplicatedSubplan(relsWithoutSameObj, tableConfig)
   }
 
+  /** Post process for the physical [[RelNode]] dag, e.g., validation or rewriting purpose. */
+  protected def postOptimize(expanded: Seq[RelNode]): Seq[RelNode]
```

Comment:
nit: give the default implementation ?

Reply:
will change it since you have the same doubts.